File: backoffice/src/shared/infrastructure/eventBus/RabbitmqEventBusConnector.py
"""
 *
 * Libraries 
 *
"""

import os
from time      import sleep
from amqpstorm import Connection
from amqpstorm import AMQPError
from amqpstorm import Channel
from injector  import singleton

"""
 *
 * Class
 *
"""

@singleton
class RabbitmqEventBusConnector:

    """
     *
     * Attributes 
     *
    """

    __connection       : Connection
    __host             : str
    __user             : str
    __password         : str
    __port             : int
    __reconnectionTime : int

    """
     *
     * Methods 
     *
    """

    def __init__( self ) -> object:
        self.__connection       = None
        self.__password         = os.getenv( 'EVENT_BUS_PASSWORD' )
        self.__user             = os.getenv( 'EVENT_BUS_USER' )
        self.__port             = int( os.getenv( 'EVENT_BUS_PORT' ) )
        self.__host             = os.getenv( 'EVENT_BUS_HOST' )
        self.__reconnectionTime = int( os.getenv( 'EVENT_BUS_RECONNECTION_TIME' ) )
    
    def getChannel( self ) -> Channel:
        return self.__getConnection().channel()

    def __getConnection( self ) -> Connection:        
        while not self.__connection:
            try:
                self.__connection = Connection(
                    username = self.__user,
                    password = self.__password,
                    hostname = self.__host,
                    port     = self.__port
                )
            except AMQPError:
                sleep( self.__reconnectionTime )
        return self.__connection

File: preparation/src/shared/infrastructure/eventBus/RabbitmqEventBusConnector.cs
/* 
 * 
 * Libraries
 *
*/

using RabbitMQ.Client;
using RabbitMQ.Client.Exceptions;

/* 
 * 
 * Class
 *
*/

namespace preparation.src.shared.infrastructure {

    public class RabbitmqEventBusConnector { 

        /* 
         * 
         * Attributes
         *
        */

        private int               reconnectionTime;
        private string            user;
        private string            password;
        private string            host;
        private string            port;
        private IConnection?      connection;
        private ConnectionFactory connectionFactory;

        /* 
         * 
         * Methods
         *
        */ 

        public RabbitmqEventBusConnector() { 
            reconnectionTime = Int32.Parse( Environment.GetEnvironmentVariable( 
                "EVENT_BUS_RECONNECTION_TIME"
            ) );
            user     = Environment.GetEnvironmentVariable( "EVENT_BUS_USER" );
            password = Environment.GetEnvironmentVariable( "EVENT_BUS_PASSWORD" );
            port     = Environment.GetEnvironmentVariable( "EVENT_BUS_PORT" );
            host     = Environment.GetEnvironmentVariable( "EVENT_BUS_HOST" );
            createConnectionFactory();
        }

        private void createConnectionFactory() {
            // Variables            
            string eventBusUri;
            // Code
            connectionFactory = new ConnectionFactory();
            eventBusUri = string.Format( 
                "amqp://{0}:{1}@{2}:{3}", 
                user, 
                password, 
                host, 
                port 
            );
            connectionFactory.Uri = new Uri( eventBusUri );
        }

        private IConnection getConnection() {            
            while( connection == null ) {
                try {                    
                    connection = connectionFactory.CreateConnection(); 
                } catch( BrokerUnreachableException ) {
                    Thread.Sleep( reconnectionTime );
                }
            }
            return connection;
        }

        public IModel getChannel() {
            return getConnection().CreateModel();
        }

    }

}
